Author:  Pavithra R <pavir@codeaurora.org>
Date:    Thu Aug 13 12:59:50 2020 +0530

This patch is propogated from the kernel 4.4 commit
56e2435c782e7cdb5c274ea012557f525d0a3b88

nat46: fix race condition in the get and release operation

when get and release the nat46 instance, it could run into race
condition, use spin_lock protect them.

Change-Id: I7a38164699a5b856f3407dae592a3d8fc82e7ffe
Signed-off-by: Pavithra R <pavir@codeaurora.org>

--- a/nat46/modules/nat46-glue.c
+++ b/nat46/modules/nat46-glue.c
@@ -18,6 +18,7 @@
 #include "nat46-glue.h"
 #include "nat46-core.h"

+static DEFINE_MUTEX(ref_lock);
 int is_valid_nat46(nat46_instance_t *nat46) {
   return (nat46 && (nat46->sig == NAT46_SIGNATURE));
 }
@@ -46,20 +47,25 @@ nat46_instance_t *alloc_nat46_instance(i

 nat46_instance_t *get_nat46_instance(struct sk_buff *sk) {
   nat46_instance_t *nat46 = netdev_nat46_instance(sk->dev);
+  mutex_lock(&ref_lock);
   if (is_valid_nat46(nat46)) {
     nat46->refcount++;
+    mutex_unlock(&ref_lock);
     return nat46;
   } else {
+    mutex_unlock(&ref_lock);
     printk("[nat46] get_nat46_instance: Could not find a valid NAT46 instance!");
     return NULL;
   }
 }

 void release_nat46_instance(nat46_instance_t *nat46) {
+  mutex_lock(&ref_lock);
   nat46->refcount--;
   if(0 == nat46->refcount) {
-    printk("[nat46] release_nat46_instance: freeing nat46 instance with %d pairs\n", nat46->npairs);
     nat46->sig = FREED_NAT46_SIGNATURE;
+    printk("[nat46] release_nat46_instance: freeing nat46 instance with %d pairs\n", nat46->npairs);
     kfree(nat46);
   }
+  mutex_unlock(&ref_lock);
 }
